lte/gateway/python/magma/pipelined/app/ipv6_solicitation.py

```diff
@@ -0,0 +1,309 @@
+        """
+        Generates the Neighbor Advertisement response packet
+        """
+        ofproto, parser = self._datapath.ofproto, self._datapath.ofproto_parser
+
+        # Only check direction OUT because direction IN doesn't need tunn info
+        if direction == Direction.OUT and (not tun_id or not tun_ipv4_dst):
+            self.logger.error("Packet missing tunnel information, can't reply")
+            return
+
+        prefix = self.get_custom_prefix(target_ipv6)
+        if not prefix:
+            self.logger.debug("Can't reply to NS for UE ip %s", target_ipv6)
+            return
+
+        pkt = packet.Packet()
+        pkt.add_protocol(
+            ethernet.ethernet(
+                dst=self.MAC_MULTICAST,
+                src=self.config.ll_addr,
+                ethertype=ether_types.ETH_TYPE_IPV6,
+            )
+        )
+        pkt.add_protocol(
+            ipv6.ipv6(
+                dst=self.DEVICE_MULTICAST,
+                src=self.config.ipv6_src,
+                nxt=in_proto.IPPROTO_ICMPV6,
+            )
+        )
+        pkt.add_protocol(
+            icmpv6.icmpv6(
+                type_=icmpv6.ND_NEIGHBOR_ADVERT,
+                data=icmpv6.nd_neighbor(
+                    dst=target_ipv6,
+                    option=icmpv6.nd_option_tla(hw_src=self.config.ll_addr)
+                ),
+            )
+        )
+        pkt.serialize()
+
+        # For NS from SGI response doesn't need tunnel information
+        actions_out = []
+        if direction == Direction.OUT:
+            actions_out.extend([
+                parser.NXActionSetTunnel(tun_id=tun_id),
+                parser.NXActionRegLoad2(dst='tun_ipv4_dst', value=tun_ipv4_dst),
+            ])
+        actions_out.append(parser.OFPActionOutput(port=output_port))
+        out = parser.OFPPacketOut(datapath=self._datapath,
+                                  buffer_id=ofproto.OFP_NO_BUFFER,
+                                  in_port=ofproto.OFPP_CONTROLLER,
+                                  actions=actions_out,
+                                  data=pkt.data)
+        self._datapath.send_msg(out)
+
+    @set_ev_cls(ofp_event.EventOFPPacketIn, MAIN_DISPATCHER)
+    def _parse_pkt_in(self, ev):
+        """
+        Process the packet in message, reply with RA/NA packets
+        """
+        msg = ev.msg
+
+        if self.tbl_num != msg.table_id:
+            # Intended for other application
+            return
+
+        in_port = ev.msg.match['in_port']
+        tun_id = None
+        tun_ipv4_src = None
+        if DIRECTION_REG not in ev.msg.match:
+            self.logger.error("Packet missing direction reg, can't reply")
+            return
+        direction = ev.msg.match[DIRECTION_REG]
+        if 'tunnel_id' in ev.msg.match:
+            tun_id = ev.msg.match['tunnel_id']
+
+        if 'tun_ipv4_src' in ev.msg.match:
+            tun_ipv4_src = ev.msg.match['tun_ipv4_src']
+
+        pkt = packet.Packet(msg.data)
+
```

Comment:
it would be good to have packet dump for debugging purpose on xmit and revcv:
self.logger.debug("pkt pkt %s", pkt.show(dump=True))

Reply:
Yeah I removed it as I thought it would be too verbose but let me add it back